components/base/wheeled/kinematics.go

```diff
@@ -0,0 +1,117 @@
+package wheeled
+
+import (
+	"bytes"
+	"context"
+
+	"github.com/golang/geo/r3"
+	"github.com/pkg/errors"
+
+	"go.viam.com/rdk/components/base"
+	"go.viam.com/rdk/config"
+	"go.viam.com/rdk/pointcloud"
+	"go.viam.com/rdk/referenceframe"
+	"go.viam.com/rdk/services/slam"
+	"go.viam.com/rdk/spatialmath"
+)
+
+type kinematicWheeledBase struct {
+	*wheeledBase
+	slam  slam.Service
+	model referenceframe.Model
+}
+
+// WrapWithKinematics takes a wheeledBase component and adds a slam service to it
+// It also adds kinematic model so that it can be controlled.
+func WrapWithKinematics(ctx context.Context, base *wheeledBase, slam slam.Service) (base.KinematicBase, error) {
+	var err error
+	kwb := &kinematicWheeledBase{
+		wheeledBase: base,
+		slam:        slam,
+	}
+	kwb.model, err = kwb.buildModel(ctx)
+	if err != nil {
+		return nil, err
+	}
+	return kwb, err
+}
+
+func (kwb *kinematicWheeledBase) ModelFrame() referenceframe.Model {
+	return kwb.model
+}
+
+func (kwb *kinematicWheeledBase) CurrentInputs(ctx context.Context) ([]referenceframe.Input, error) {
+	// TODO(RSDK-2311): complete the implementation
+	return []referenceframe.Input{}, errors.New("not implemented yet")
+}
+
+func (kwb *kinematicWheeledBase) GoToInputs(ctx context.Context, goal []referenceframe.Input) error {
+	// TODO(RSDK-2311): complete the implementation
+	return errors.New("not implemented yet")
+}
+
+// ModelFrame builds the kinematic model associated with the kinematicWheeledBase
+// Note that this model is not intended to be registered in the frame system
+func (kwb *kinematicWheeledBase) buildModel(ctx context.Context) (referenceframe.Model, error) {
+	if kwb.collisionGeometry == nil {
+		return nil, errors.New("cannot create model for base with no collision geometry")
+	}
+
+	// get the limits of the SLAM map to set as the extents of the frame
+	data, err := slam.GetPointCloudMapFull(ctx, kwb.slam, "")
+	if err != nil {
+		return nil, err
+	}
+	pcd, err := pointcloud.ReadPCD(bytes.NewReader(data))
```

Comment:
For large pointclouds, this could be an expensive operation. We could either just parse the metadata here (by exposing this function: https://github.com/viamrobotics/rdk/blob/23b8fcf5fe7866cd18c349165707066cf435da2e/pointcloud/pointcloud_file.go#L667) or read the pcd to an octree here, and save the octree for later use in the Move() call.

Reply:
thanks I was not aware of this function! I will use the PCD MetaData for the time being, since I'm not sure how we're going to want to access/use the octrees themselves

Reply:
Sounds good! We'll need to make getPCDMetaData a public function.

Reply:
I pushed something that does this, let me know what you think

Reply:
This looks good to me! You should add Bijan to reviewers for pointcloud changes.

Reply:
@nicksanford , see this thread